framework/test/util/ConcurrencyTestUtil.hpp

```diff
@@ -0,0 +1,64 @@
+
+
+#ifndef ConcurrencyTestUtil_hpp
+#define ConcurrencyTestUtil_hpp
+
+#include <algorithm>
+#include <functional>
+#include <future>
+#include <vector>
+
+namespace cppmicroservices {
+namespace detail {
+namespace test {
+/**
+ * Util method to determine if a future object is ready
+ */
+template<typename T>
+bool is_ready(T&& futObj)
+{
+  return (futObj.wait_for(std::chrono::seconds::zero()) ==
+          std::future_status::ready);
+}
+
+/**
+ * Utility method for concurrently running a function
+ */
+template<class R, class... Args>
+std::vector<R> ConcurrentInvoke(std::function<R(Args...)> func)
+{
+  // test concurrent calls to enable and disable from multiple threads
+  std::vector<R> returnVals;
+  std::promise<void> go;
+  std::shared_future<void> ready(go.get_future());
+  size_t numCalls = std::max(64u, std::thread::hardware_concurrency());
+  std::vector<std::promise<void>> readies(numCalls);
+
+  // TODO: use std::barrier when available
+  std::vector<std::future<R>> enable_async_futs(numCalls);
+  for (size_t i = 0; i < numCalls; ++i) {
+    enable_async_futs[i] =
+      std::async(std::launch::async, [&readies, i, &ready, &func]() {
+        readies[i].set_value();
```

Comment:
func() is executed before the readies[i].set_value call. I think it should be executed after the func() call. Otherwise the readies[i].get_future().wait() in the next for loop could potentially be satisfied and move on before the last function has finished executing.

Reply:
This is functioning as designed from what I can tell. The `readies` vector is meant to synchronize all the threads to right before `func()` will be called. Then once `go.set_value()` executes, all the threads race to execute `func()`. `ready.wait()` stops `func()` from being executed prematurely.